feat: reuse runtime services and wrappers between XBlocks [FC-0026]

Agrendalath opened the pull request:
## Description

The goal of FC-0026 is to remove the XBlock-specific handling from the prepare_runtime_for_user function. This part handles the final step, which is introducing runtime service reusability.

In a single request, each XBlocks use the same runtime instance. Therefore, when the services are initialized for the user in the course, these services are overridden[^1], so all XBlock-specific data is lost. In previous FC-0026 PRs, we have removed the block-specific code from the `prepare_runtime_for_user` function. This way, it no longer matters which XBlock bound the runtime to the user. This PR ensures that the runtime binding happens only once for each user[^2].

[^1]: You can see [this comment](https://github.com/openedx/edx-platform/pull/32368#issuecomment-1594679000), which explains the fragile behavior related to the runtime initialization order.
[^2]: There are two use cases for this:
    1. While rebinding the user in the LTI XBlock.
    1. In the Instructor Dashboard tasks - when a Problem Block is rescored for multiple users.

## Supporting information

Private-ref: [BB-7448](https://tasks.opencraft.com/browse/BB-7448)

## Testing instructions

Ensure that XBlocks work correctly in the LMS, Preview, and Studio. Check that importing and exporting the course works.

## Other notes
Please ignore the first two commits, as they come from #32357.

openedx-webhooks commented:
Thanks for the pull request, @Agrendalath!

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.



<!-- comment:welcome-core-committer -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

Agrendalath committed:
refactor: reuse services and wrappers between XBlocks

Agrendalath commented on lms/djangoapps/courseware/tests/test_block_render.py:
```diff
@@ -950,7 +950,7 @@ def test_will_recheck_access_handler_attribute(self, handler, will_recheck_acces
 
         render.handle_xblock_callback(request, str(course.id), usage_id, handler)
-        assert mock_get_block.call_count == 1
+        assert mock_get_block.call_count == 2
```
This has increased because we renamed `get_block_for_descriptor_internal` to `get_block_for_descriptor`.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -582,84 +651,11 @@ def inner_get_block(block):
-            block.has_access_error = bool(caller_will_handle_access_error)
-            return block
-        return None
-    return block
```
I moved this code to `get_block_for_descriptor`. I find this change a bit controversial - allowing `None` values in positional arguments is inelegant.

However, I had to untangle the dependencies between `get_block_for_descriptor` and `get_block_for_descriptor_internal` whenever I read this code. The latter exists primarily because we can use it in the `inner_get_block` and Instructor Dashboard tasks without the request object. Unifying these two functions makes this code (subjectively) easier to understand. We won't need to jump between four functions (`get_block_for_descriptor`, `get_block_for_descriptor_internal`, `prepare_runtime_for_user`, `inner_get_block`) to understand why they are calling each other.

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -1536,11 +1535,8 @@ def service(self, block, service_name):
-            # getting the service from parent module. making sure of block service declarations.
-            service = super().service(block=block, service_name=service_name)
+        # Getting the service from parent module. making sure of block service declarations.
+        service = super().service(block=block, service_name=service_name)
```
We don't need `runtime._runtime_services` anymore. Within a single request, each XBlock uses the same instance of the runtime, so we can safely rely on the `XBlock` library to check the service declaration:
https://github.com/openedx/XBlock/blob/7b8bdeafbabbbc96f11dfbceaf53c89bd416ea6f/xblock/runtime.py#L1120-L1126.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +370,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
+            student_kvs = MasqueradingKeyValueStore(student_kvs, request.session)
+        student_data = KvsFieldData(student_kvs)
+
+    # The runtime is already shared between XBlocks. If there are no wrappers, it is the first initialization.
```
This. Each XBlock in a single request uses the same runtime instance, so we don't need any cache. This was why the services were overriding each other with each XBlock initialization.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -482,16 +561,6 @@ def inner_get_block(block):
             request_token=request_token,
         ))
 
-    # HACK: The following test fails when we do not access this attribute.
-    #  lms/djangoapps/courseware/tests/test_views.py::TestRenderXBlock::test_success_enrolled_staff
-    #  This happens because accessing `field_data` caches the XBlock's parents through the inheritance mixin.
-    #  If we do this operation before assigning `inner_get_block` to the `get_block_for_descriptor` attribute, these
-    #  parents will be initialized without binding the user data (that's how the `get_block` works in `x_module.py`).
-    #  If we retrieve the parents after this operation (e.g., in the `enclosing_sequence_for_gating_checks` function),
-    #  they will have their runtimes initialized, which can lead to an altered behavior of the XBlock-specific
-    #  parts of other runtimes. We will keep this workaround until we remove all XBlock-specific code from here.
-    block.static_asset_path  # pylint: disable=pointless-statement
```
Fixed by dc3fd30. [This conversation](https://github.com/openedx/edx-platform/pull/8240#discussion_r31749841) explains why `disable_staff_debug_info` was added here. The `render_xblock` view is used in two places:
1. In the Learning MFE (as an iframe).
2. In the LTI context where Open edX is the LTI Provider.

It looks like the "staff debug info" button has been working correctly in the MFE by a pure accident - this parameter was overridden when the view initialized parent XBlocks.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +370,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
-        disable_staff_debug_info=disable_staff_debug_info,
-        course=course,
-        will_recheck_access=will_recheck_access,
+    if request:
+        track_function = track_function or make_track_function(request)
+        user_location = user_location or getattr(request, 'session', {}).get('country_code')
+        request_token = request_token or xblock_request_token(request)
+
+    if not student_data:
+        student_kvs = DjangoKeyValueStore(field_data_cache)
+        if is_masquerading_as_specific_student(user, course_key):
+            student_kvs = MasqueradingKeyValueStore(student_kvs, request.session)
+        student_data = KvsFieldData(student_kvs)
```
Added because of the `get_block_for_descriptor_internal` -> `get_block_for_descriptor` merge.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +370,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
+        student_data = KvsFieldData(student_kvs)
+
+    # The runtime is already shared between XBlocks. If there are no wrappers, it is the first initialization.
+    should_recreate_runtime = not block.runtime.wrappers
```
We are not checking services because some of them are initialized by the `Runtime` class from the `XBlock` library.

Agrendalath commented on xmodule/services.py:
```diff
@@ -201,10 +195,11 @@ def rebind_noauth_module_to_user(self, block, real_user):
             course = modulestore().get_course(course_key=self.course_id)
 
-        inner_student_data = self._ref["prepare_runtime_for_user"](
+        from lms.djangoapps.courseware.block_render import prepare_runtime_for_user
```
I don't know why we passed the function to this service - the local import is working fine.

ormsbee replied:
My guesses:

1. We were trying to reduce the number of direct dependencies `xmodule` had on `lms.*`.
2. We used to pass different methods for LMS/Studio/Preview here?

Agrendalath replied:
@ormsbee, neither of these. We added this in https://github.com/openedx/edx-platform/pull/30320. I might have forgotten to suggest changing it there.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +372,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
-        disable_staff_debug_info=False,
-        course=None,
-        will_recheck_access=False,
+        runtime: Runtime,
```
We removed all XBlock-specific code from this function, so passing the runtime directly (instead of the `block`) is better.

Agrendalath commented:
@0x29a, @bradenmacdonald, this is ready for review.

bradenmacdonald commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +372,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
+        if user_service := block.runtime._services.get('user'):  # pylint: disable=protected-access
+            # Check the user ID bound to the runtime. This operation can run often for complex course structures, so we
+            # are accessing the protected attribute of the user service to reduce the number of queries.
+            should_recreate_runtime = user.id != user_service._django_user.id  # pylint: disable=protected-access
```
What happens here if the user is masquerading? Is `user.id` and/or `user_service._django_user.id` the "real" user ID or the masquerading user ID?

Agrendalath replied:
When a user is masquerading, the "real" user is under the `user.real_user`:
https://github.com/openedx/edx-platform/blob/c80fba689a36aa35489ded8018eb997c4ef4e8a1/lms/djangoapps/courseware/masquerade.py#L205-L244

The user service contains our masquerading target:
https://github.com/openedx/edx-platform/blob/27969b3e1e82edf35ac92a2eeae99ad15d6bf949/lms/djangoapps/courseware/block_render.py#L603-L613

Therefore, masquerading would require rebinding the runtime only if the runtime initialization happened from the "real" user (though I'm unsure if this ever happens). Once the runtime is bound to our masquerading target, the user is detected correctly, as the user service is initialized with the `user` object.

bradenmacdonald commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -364,60 +372,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
+            log.exception('Non-integer %r passed as position.', position)
+            position = None
+
+    block.runtime.set('position', position)
```
What is this `position` parameter? I'm not familiar with it.

Agrendalath replied:
As far as I understand, this is used by the `SequenceBlock`/subsection to determine which tab should be active:
https://github.com/openedx/edx-platform/blob/0b455e0336fa54199fd992c2ce83a81c07dd07c2/xmodule/seq_block.py#L335-L363

This is no longer used in the learning MFE, but the Preview still uses the old experience. I tried removing this, but it caused some errors. It should be safe to remove when we completely remove the legacy experience. For now, I would leave it as-is.

bradenmacdonald replied:
I see. Could you perhaps add that information as a comment into the function?

Agrendalath replied:
Sure, I added it in d7b898d. I added this information to the `get_block` function since it contains the description of these parameters. Let me know if you had another function in mind.

bradenmacdonald commented on xmodule/services.py:
```diff
@@ -201,10 +195,11 @@ def rebind_noauth_module_to_user(self, block, real_user):
             course = modulestore().get_course(course_key=self.course_id)
 
-        self._ref["prepare_runtime_for_user"](
+        from lms.djangoapps.courseware.block_render import prepare_runtime_for_user
```
`RebindUserService` is only used in the LMS right? I almost wonder if we should move this code out of `xmodule` and into `lms`. But no change needed.

Agrendalath replied:
Yes, we only use it in LMS. However, we also instantiate it in the `XBlockRuntime`, which is in the `openedx.core.djangoapps.xblock`. I don't have enough context about this runtime to determine if we need it there:

https://github.com/openedx/edx-platform/blob/27969b3e1e82edf35ac92a2eeae99ad15d6bf949/openedx/core/djangoapps/xblock/runtime/runtime.py#L265-L274

bradenmacdonald replied:
That runtime is currently used for "v2" content libraries and was also used for LabXchange where learners could "learn" (LMS mode) directly from components in a content library. I don't think it's used for LabXchange anymore. So the rebind service there is probably not used at the moment but will be in the future.

bradenmacdonald reviewed (approved):
👏🏻 wow, I am so happy to see this.

Again, I'll let @0x29a do the regression testing etc. I've just read through the code and I have to say, I really like it. Much cleaner and easier to understand. Hopefully more efficient and less buggy too :)

0x29a reviewed (approved):
:+1:

<details> 
  <summary>Note: this error appears when deleting a block from a unit in Studio. I verified that it happens without the changes from this PR and for courses created earlier, but anyway decided to share, just in case.</summary>

```python
2023-06-26 11:42:45,652 INFO 1058 [tracking] [user 3] [ip 172.21.0.1] logger.py:41 - {"name": "/xblock/block-v1:Test+Test+Test+type@openassessment+block@878d7bca725b405ea5ec19ec28059df6", "context": {"user_id": 3, "path": "/xblock/block-v1:Test+Test+Test+type@openassessment+block@878d7bca725b405ea5ec19ec28059df6", "course_id": "", "org_id": "", "enterprise_uuid": ""}, "username": "edx", "session": "74b02955cff3654d45e4205f713fe972", "ip": "172.21.0.1", "agent": "Mozilla/5.0 (Windows NT 10.0; rv:114.0) Gecko/20100101 Firefox/114.0", "host": "localhost:18010", "re
ferer": "http://localhost:18010/container/block-v1:Test+Test+Test+type@vertical+block@552e0a3f2484401192e53b16cda50d9a?action=new", "accept_language": "en-US,en;q=0.5", "event": "{\"GET\": {}, \"POST\": {}}", "time": "2023-06-26T11:42:45.651868+00:00", "event_type": "/xblock/block-v1:Test+Test+Test+type@openassessment+block@878d7bca725b405ea5ec19ec28059df6", "event_source": "server", "page": null}
Error calling handle_item_deleted in Signal.send_robust() (BlockKey(type='openassessment', id='878d7bca725b405ea5ec19ec28059df6'))                                                                                                                                                          Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 212, in send_robust                                                                                                                                                                       response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/signals/handlers.py", line 254, in handle_item_deleted                                                                                                                                                                         deleted_block = modulestore().get_item(usage_key)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 84, in inner                                                                                                                                                                                                           retval = func(field_decorator=strip_key_collection, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 250, in get_item                                                                                                                                                                                                   
    return store.get_item(usage_key, depth, **kwargs)                                                                                                                                                                                                                                         File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split_draft.py", line 283, in get_item
    return super().get_item(usage_key, depth=depth, **kwargs)                                                                                                                                                                                                                               
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 1182, in get_item                                                                                                                                                                                          items = self._load_items(course, [BlockKey.from_usage_key(usage_key)], depth, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 779, in _load_items                                                                                                                                                                                    
    return [runtime.load_item(block_key, course_entry, **kwargs) for block_key in block_keys]                                                                                                                                                                                               
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 779, in <listcomp>                                                                                                                                                                                     
    return [runtime.load_item(block_key, course_entry, **kwargs) for block_key in block_keys]                                                                                                                                                                                               
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/caching_descriptor_system.py", line 122, in _load_item              
    block_data = self.get_module_data(block_key, course_key)                                                                                                                                                                                                                                  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/caching_descriptor_system.py", line 149, in get_module_data                                                                                                                                                                raise ItemNotFoundError(block_key)                                                                                                                                                                                                                                                      xmodule.modulestore.exceptions.ItemNotFoundError: BlockKey(type='openassessment', id='878d7bca725b405ea5ec19ec28059df6')        
2023-06-26 11:42:45,661 ERROR 1058 [django.dispatch] [user 3] [ip 172.21.0.1] dispatcher.py:214 - Error calling handle_item_deleted in Signal.send_robust() (BlockKey(type='openassessment', id='878d7bca725b405ea5ec19ec28059df6'))
Traceback (most recent call last):                                                                                                                                                                                                                                                            File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 212, in send_robust                                                                                                                                                                       response = receiver(signal=self, sender=sender, **named)                                                                                                                                                                                                                                
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/signals/handlers.py", line 254, in handle_item_deleted
    deleted_block = modulestore().get_item(usage_key)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 84, in inner
    retval = func(field_decorator=strip_key_collection, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 250, in get_item
    return store.get_item(usage_key, depth, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split_draft.py", line 283, in get_item
    return super().get_item(usage_key, depth=depth, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 1182, in get_item
    items = self._load_items(course, [BlockKey.from_usage_key(usage_key)], depth, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 779, in _load_items
    return [runtime.load_item(block_key, course_entry, **kwargs) for block_key in block_keys]
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 779, in <listcomp>
    return [runtime.load_item(block_key, course_entry, **kwargs) for block_key in block_keys]
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/caching_descriptor_system.py", line 122, in _load_item
    block_data = self.get_module_data(block_key, course_key)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/caching_descriptor_system.py", line 149, in get_module_data
    raise ItemNotFoundError(block_key)
xmodule.modulestore.exceptions.ItemNotFoundError: BlockKey(type='openassessment', id='878d7bca725b405ea5ec19ec28059df6')
```

</details>

- [x] I did plenty of regression tests.
- [x] I read through the changes.
- [x] Includes comments where needed.

Agrendalath commented:
@ormsbee, would you like to take a look at this (and #32357) before we schedule the merge?

ormsbee reviewed (approved):
This PR brings me so much joy. 😄 Great stuff!

Agrendalath committed:
refactor: remove redundant masquerade handling from runtime initialization

This part was added in 3d7246ecc81e3cb9e040b41f49c17e1da7e00b0a.
Then, 0dd4978f37b8d6d581e687a07ee5025b6693727d refactored the approach, but
did not remove these lines.

ormsbee commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -97,6 +98,13 @@
 from common.djangoapps.xblock_django.user_service import DjangoXBlockUserService
 from openedx.core.lib.cache_utils import CacheService
 
+if TYPE_CHECKING:
+    from rest_framework.request import Request
+    from xblock.core import XBlock
+    from xblock.runtime import Runtime
+
+    from xmodule.course_block import CourseBlock
```
Why are these conditional imports?

Agrendalath replied:
@ormsbee, [typing.TYPE_CHECKING](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) (defined in [PEP-484](https://peps.python.org/pep-0484/)) is `False` at runtime. It should only be `True` during static analysis (usually in an IDE or type checker). Thanks to this, and the fact we don't evaluate annotations in function definitions ([PEP-563](https://peps.python.org/pep-0563/)), we:
1. Reduce the number of imports. These imports are not heavy, but we're not instantiating these classes, so we don't need them at runtime.
2. Avoid potential circular imports at runtime.

openedx-webhooks commented:
@Agrendalath 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/32420&entry.1484655318=https://github.com/Agrendalath&entry.752974735=2023-06-12+11:50&entry.1917517419=2023-06-30+13:29&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
